### PR TITLE
feat(denommetadata): register IBC denom on transfer 

### LINKED ISSUE #955: feat(ibc transfer): Register IBC denom on transfer
On IBC transfer from Rollapp to the Hub, register any missing denom metadata on the Hub.

https://www.notion.so/dymension/ADR-x-IBC-Denom-Metadata-Transfer-From-Rollapp-to-Hub-54e74e50adeb4d77b1f8777c05a73390?pvs=4

### PR DESCRIPTION
**Pull request by zale144**
## Description

----

Closes #955 

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow-up issues.

PR review checkboxes:

I have...

- [ ]  Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ]  Targeted PR against the correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to the GitHub issue with discussion and accepted design
- [ ]  Targets only one GitHub issue
- [ ]  Wrote unit and integration tests
- [ ]  Wrote relevant migration scripts if necessary
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)
- [ ]  Updated the scripts for local run, e.g genesis_config_commands.sh if the PR changes parameters
- [ ]  Add an issue in the [e2e-tests repo](https://github.com/dymensionxyz/e2e-tests) if necessary

SDK Checklist
- [ ] Import/Export Genesis
- [ ] Registered Invariants
- [ ] Registered Events
- [ ] Updated openapi.yaml
- [ ] No usage of go `map`
- [ ] No usage of `time.Now()`
- [ ] Used fixed point arithmetic and not float arithmetic
- [ ] Avoid panicking in Begin/End block as much as possible
- [ ] No unexpected math Overflow
- [ ] Used `sendCoin` and not `SendCoins`
- [ ] Out-of-block compute is bounded
- [ ] No serialized ID at the end of store keys
- [ ] UInt to byte conversion should use BigEndian

Full security checklist [here](https://www.faulttolerant.xyz/2024-01-16-cosmos-security-1/)

----;

For Reviewer:

- [ ]  Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  Confirmed all author checklist items have been addressed

---;

After reviewer approval:

- [ ]  In case the PR targets the main branch, PR should not be squash merge in order to keep meaningful git history.
- [ ]  In case the PR targets a release branch, PR must be rebased.

### CODE REVIEW

**Comment by mtsitrin** [comment]
UT fails

File: utils/ibc.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+	return GetForeignDenomTrace(channelId, denom).IBCDenom()
+}
+
+func GetForeignDenomTrace(channelId string, denom string) transfertypes.DenomTrace {
+	sourcePrefix := transfertypes.GetDenomPrefix(ibcPort, channelId)
+	// NOTE: sourcePrefix contains the trailing "/"
+	prefixedDenom := sourcePrefix + denom
+	// construct the denomination trace from the full raw denomination
+	return transfertypes.ParseDenomTrace(prefixedDenom)
+}
```
Comment (danwt) [inline utils/ibc.go:21]: there is already a ibc utils package

File: x/rollapp/keeper/util.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+	}
+
+	if rollapp == nil {
+		return "", packet, nil
+	}
+
+	return rollapp.RollappId, packet, nil
+}
+
+// ExtractRollappFromChannel extracts the rollapp from the IBC port and channel.
```
Comment (danwt) [inline x/rollapp/keeper/util.go:40]: should use https://github.com/dymensionxyz/dymension/blob/4565f34fe8a8d2cd9e34a42bd6e3cf71ecbfecd9/x/rollapp/keeper/authenticate_packet.go#L21
all the other middlewares use the pattern, it's easy to copy

File: x/denommetadata/types/errors.go [hunk 3]
```diff
@@ -8,5 +8,7 @@ import (
 
 // x/denommetadata module sentinel errors
 var (
-	ErrUnknownRequest = errorsmod.Register(ModuleName, 1002, "unknown request")
+	ErrDenomAlreadyExists = errorsmod.Register(ModuleName, 1000, "denom metadata is already registered")
+	ErrUnknownRequest     = errorsmod.Register(ModuleName, 1002, "unknown request")
+	ErrRollappNotFound    = errorsmod.Register(ModuleName, 1003, "rollapp not found")
 )
```
Comment (danwt) [inline x/denommetadata/types/errors.go:14]: we should use gerrc as a basis of errors now

https://github.com/dymensionxyz/gerr-cosmos/blob/main/gerrc/errors.go#L26-L41

File: x/denommetadata/types/packet_metadata.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	DenomMetadata *types.Metadata `json:"denom_metadata,omitempty"`
+}
+
+func (p MemoData) ValidateBasic() error {
+	return p.DenomMetadata.Validate()
+}
+
+const memoObjectKeyDenomMetadata = "denom_metadata"
+
+var ErrMemoDenomMetadataAlreadyExists = errorsmod.Wrapf(errortypes.ErrUnauthorized, "'denom_metadata' already exists in memo")
```
Comment (danwt) [inline x/denommetadata/types/packet_metadata.go:22]: gerrc already exist

**Review by danwt — CHANGES_REQUESTED** [review]
Please, instead of reintroducing all the old stuff, just move the ibc module and ics4 wrapper from transferinject and put your changes in there

If you look at main, currently, all the middlewares are very consistent in their pattern and use the authenticate packet utils

https://github.com/dymensionxyz/dymension/blob/4565f34fe8a8d2cd9e34a42bd6e3cf71ecbfecd9/x/rollapp/keeper/authenticate_packet.go#L21

It's all been dried out

Also please for errors, now prefer to use our gerr cosmos package and wrap them instead of defining new errors a lot. (You can still define new errors, but it's rare to need to do so)

https://pkg.go.dev/github.com/dymensionxyz/gerr-cosmos@v0.1.2/gerrc#pkg-variables

File: x/denommetadata/ibc_middleware.go [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+
+	rollappKeeper types.RollappKeeper
+	bankKeeper    types.BankKeeper
+}
+
+// NewIBCSendMiddleware creates a new ICS4Wrapper.
+// It intercepts outgoing IBC packets and adds token metadata to the memo if the rollapp doesn't have it.
+// This is a solution for adding token metadata to fungible tokens transferred over IBC,
+// targeted at rollapps that don't have the token metadata for the token being transferred.
+// More info here: https://www.notion.so/dymension/ADR-x-IBC-Denom-Metadata-Transfer-From-Hub-to-Rollapp-d3791f524ac849a9a3eb44d17968a30b
```
Comment (danwt) [inline x/denommetadata/ibc_middleware.go:172]: please put this on the type not the `New`

File: x/denommetadata/ibc_middleware.go [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+
+	// get the denom metadata from the bank keeper, if it doesn't exist, move on to the next middleware in the chain
+	denomMetadata, ok := m.bankKeeper.GetDenomMetaData(ctx, packet.Denom)
+	if !ok {
+		return m.ICS4Wrapper.SendPacket(ctx, chanCap, destinationPort, destinationChannel, timeoutHeight, timeoutTimestamp, data)
+	}
+
+	packet.Memo, err = types.AddDenomMetadataToMemo(packet.Memo, denomMetadata)
+	if err != nil {
+		return 0, errorsmod.Wrapf(errortypes.ErrUnauthorized, "add denom metadata to memo: %s", err.Error())
```
Comment (danwt) [inline x/denommetadata/ibc_middleware.go:235]: why unauthorized?

File: x/denommetadata/keeper/keeper.go [hunk 7]
```diff
@@ -46,7 +47,7 @@ func (k *Keeper) CreateDenomMetadata(ctx sdk.Context, metadata banktypes.Metadat
 
 // UpdateDenomMetadata returns the denommetadata of the specified denom
 func (k *Keeper) UpdateDenomMetadata(ctx sdk.Context, metadata banktypes.Metadata) error {
-	found := k.HasDenomMetadata(ctx, metadata.Base)
+	_, found := k.bankKeeper.GetDenomMetaData(ctx, metadata.Base)
 	if !found {
```
Comment (danwt) [inline x/denommetadata/keeper/keeper.go:51]: why this change?
if denommetadat keeper.HasDenomMetadata changes impl then this will not be consistent

File: x/denommetadata/proposal_handler.go [hunk 8]
```diff
@@ -16,7 +18,7 @@ func NewDenomMetadataProposalHandler(k *keeper.Keeper) govtypes.Handler {
 		case *types.UpdateDenomMetadataProposal:
 			return HandleUpdateDenomMetadataProposal(ctx, k, c)
 		default:
-			return errorsmod.Wrapf(types.ErrUnknownRequest, "unrecognized denommetadata proposal content type: %T", c)
+			return errorsmod.Wrapf(gerrc.ErrUnknown, "unrecognized denommetadata proposal content type: %T", c)
```
Comment (danwt) [inline x/denommetadata/proposal_handler.go:21]: `			return errorsmod.WithType(gerrc.ErrInvalidArgument, c)
`

Comment (danwt) [inline x/denommetadata/proposal_handler.go:21]:
```suggestion
			return errorsmod.WithType(gerrc.ErrInvalidArgument, c)
```

**Review by danwt — CHANGES_REQUESTED** [review]
Looks good man

Please see my comments on the rdk pr and then read this, because they are similar.

Basically, please name the structs IBCModule or ICS4Wrapper, not middleware, because they aren't actually middleware

but basically gj

**Review by danwt — DISMISSED** [review]
great job

File: x/denommetadata/ibc_middleware.go [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+
+	if err = im.keeper.CreateDenomMetadata(ctx, *dm); err != nil {
+		if errorsmod.IsOf(err, gerrc.ErrAlreadyExist) {
+			return im.IBCModule.OnRecvPacket(ctx, packet, relayer)
+		}
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
+	if !im.transferKeeper.HasDenomTrace(ctx, denomTrace.Hash()) {
+		im.transferKeeper.SetDenomTrace(ctx, denomTrace)
```
Comment (mtsitrin) [inline x/denommetadata/ibc_middleware.go:96]: why u need to set it for the transferKeeper? isn't it gonna be created anyway in the inner `OnRecvPacket`?

File: x/denommetadata/ibc_middleware.go [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+// It will intercept an incoming packet and check if the denom metadata exists.
+// If it does not, it will register the denom metadata.
+// The handler will expect a 'denom_metadata' object in the memo field of the packet.
+// If the memo is not an object, or does not contain the metadata, it moves on to the next handler.
+func (im IBCModule) OnRecvPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	relayer sdk.AccAddress,
+) exported.Acknowledgement {
+	transferData, err := im.rollappKeeper.GetValidTransfer(ctx, packet.Data, packet.DestinationPort, packet.DestinationChannel)
```
Comment (mtsitrin) [inline x/denommetadata/ibc_middleware.go:56]: probably should add

```
	if commontypes.SkipRollappMiddleware(ctx) {
		return w.IBCModule.OnRecvPacket(ctx, packet, relayer)
	}
```

File: x/denommetadata/ibc_middleware.go [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+	// at this point it's safe to assume that we are not handling a native token of the rollapp
+	denomTrace := utilsibc.GetForeignDenomTrace(packet.GetDestChannel(), packetData.Denom)
+	ibcDenom := denomTrace.IBCDenom()
+
+	dm := types.ParsePacketMetadata(packetData.Memo)
+	if dm == nil {
+		return im.IBCModule.OnRecvPacket(ctx, packet, relayer)
+	}
+
+	if err = dm.Validate(); err != nil {
```
Comment (mtsitrin) [inline x/denommetadata/ibc_middleware.go:76]: don't we want to validate the memo base is the same as the actual `packetData.Denom`?

File: x/denommetadata/ibc_middleware.go [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+	}
+
+	rollapp, packetData := transferData.Rollapp, transferData.FungibleTokenPacketData
+
+	if packetData.Memo == "" {
+		return im.IBCModule.OnRecvPacket(ctx, packet, relayer)
+	}
+
+	// at this point it's safe to assume that we are not handling a native token of the rollapp
+	denomTrace := utilsibc.GetForeignDenomTrace(packet.GetDestChannel(), packetData.Denom)
```
Comment (mtsitrin) [inline x/denommetadata/ibc_middleware.go:68]: need to validate it's not token originating from the hub?
( `if transfertypes.ReceiverChainIsSource` ...)

File: x/denommetadata/ibc_middleware.go [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+			return im.IBCModule.OnRecvPacket(ctx, packet, relayer)
+		}
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
+	if !im.transferKeeper.HasDenomTrace(ctx, denomTrace.Hash()) {
+		im.transferKeeper.SetDenomTrace(ctx, denomTrace)
+	}
+
+	if !Contains(rollapp.RegisteredDenoms, dm.Base) {
```
Comment (mtsitrin) [inline x/denommetadata/ibc_middleware.go:99]: isn't the `rollapp.RegisteredDenoms` used for hub->RA denoms registration?
the tokens **from** RA->Hub supposed to be for RA native tokens.
not sure why u need maintain a list of denoms, as u already set it in the bank module

**Review by mtsitrin — CHANGES_REQUESTED** [review]

File: x/denommetadata/ibc_middleware.go (hunk 12 above)
Comment (zale144) [reply x/denommetadata/ibc_middleware.go:68]: it's already checked on the sender side (RA): https://github.com/dymensionxyz/dymension-rdk/blob/90caf4aa392f3d9f05abc0d4febeefcb83c33756/x/denommetadata/ibc_middleware.go#L81

File: x/denommetadata/ibc_middleware.go (hunk 13 above)
Comment (zale144) [reply x/denommetadata/ibc_middleware.go:99]: if we have a situation where the rollapp first registers a denom on the hub, (RA -> HUB), later if the hub wants to send a transfer with the same denom to the rollapp (HUB -> RA), in order to prevent the same denom metadata to be included in the transfer to the rollapp, we add it to `rollapp.RegisteredDenoms` on the hub when handling `OnRecvPacket`. 

wdym by 

> the tokens from RA->Hub supposed to be for RA native tokens.

why can't the rollapp register a non-native denom to the hub? The native RA denom is already registered during the "genesis event", or whatever we call it these days.

Comment (mtsitrin) [reply x/denommetadata/ibc_middleware.go:101]: `in order to prevent the same denom metadata to be included in the transfer to the rollapp`
the hub shouldn't try and register denoms orignating from the rollapp anyway

Comment (mtsitrin) [reply x/denommetadata/ibc_middleware.go:101]: `why can't the rollapp register a non-native denom to the hub?`
the rollapp can't register `ibc/` denoms. i.e only native denoms

**Review by mtsitrin — APPROVED** [review]
